crates/holochain_integrity_types/src/chain.rs

```diff
@@ -16,87 +16,96 @@ mod test;
 /// Filter source chain items.
 /// Starting from some chain position given as an [`ActionHash`]
 /// the chain is walked backwards to genesis.
-/// The filter can stop early by specifying the number of
-/// chain items to take and / or an [`ActionHash`] to consume until.
+/// The filter can stop early by specifying stop conditions:
+/// A maximum number of items is reached, a given [`ActionHash`] is found, or
+/// a given timestamp has been passed.
+/// Multiple `[`ActionHash`] can be provided. The filter will stop at and take the first one found.
+/// When providing a Timestamp, the filter will stop at and take the last item that has a
+/// Timestamp younger than the provided one. In the edge case of multiple actions having the same
+/// Timestamp as the stop condition, only one of the actions will be included.
 pub struct ChainFilter<H: Eq + Ord + std::hash::Hash = ActionHash> {
     /// The starting position of the filter.
     pub chain_top: H,
-    /// The filters that have been applied.
-    /// Defaults to [`ChainFilters::ToGenesis`].
-    pub filters: ChainFilters<H>,
+    /// The stop conditions used by this filter.
+    /// Defaults to [`StopConditions::ToGenesis`].
+    pub stop_conditions: StopConditions<H>,
     /// Should the query return any entries that are
     /// cached at the agent activity to save network hops.
     pub include_cached_entries: bool,
 }
 
 #[derive(Serialize, Deserialize, Debug, Eq, Clone)]
-/// Specify which [`Action`](crate::action::Action)s to allow through
-/// this filter.
-pub enum ChainFilters<H: Eq + Ord + std::hash::Hash = ActionHash> {
+/// Specify when to stop walking down the chain.
+pub enum StopConditions<H: Eq + Ord + std::hash::Hash = ActionHash> {
     /// Allow all up to genesis.
     ToGenesis,
-    /// Take this many (inclusive of the starting position).
+    /// Take this many items (inclusive of the starting position).
     Take(u32),
+    /// Continue until some timestamp is passed.
```

Comment:
The chain is walked backwards, so this should be something like "All actions since the given timestamp"?